### PR TITLE
Set minimum length on passwords

### LINKED ISSUE #19: Set minimum length on passwords
Right now we allow people to set 0 length passwords. We should instead require a minimum of 6-8 characters for their passwords when user is created. This should be a configurable traitlet.

We shouldn't enforce this for current passwords though, since that can lock users out of their hubs.

### PR DESCRIPTION
**Pull request by GeorgianaElena**
@yuvipanda, I tried to add the option to have a minimum password length as specified [here](https://github.com/jupyterhub/firstuseauthenticator/issues/19) and a test for it.

Please let me know what you think :eyes: 

EDIT: Closes #19

### CODE REVIEW

File: firstuseauthenticator/firstuseauthenticator.py [hunk 1]
```diff
@@ -108,6 +122,13 @@ def authenticate(self, handler, data):
                 return None
 
         password = data['password']
+        # Don't enforce password length requirement on existing users, since that can
+        # lock users out of their hubs.
+        if not self._validate_password(password) and not self._user_exists(username):
+            self.log.error('Password too short! \
```
Comment (yuvipanda) [inline firstuseauthenticator/firstuseauthenticator.py]: Does this display the error to the user? If not, is there an easy way for us to do that? That'd be great...

File: firstuseauthenticator/firstuseauthenticator.py [hunk 2]
```diff
@@ -134,6 +155,11 @@ def reset_password(self, username, new_password):
         """
         This allow to change password of a logged user.
         """
+        if not self._validate_password(password):
+            self.log.error('Password too short! \
```
Comment (yuvipanda) [inline firstuseauthenticator/firstuseauthenticator.py]: Same here - does this message display to the user?

File: firstuseauthenticator/firstuseauthenticator.py (hunk 1 above)
Comment (GeorgianaElena) [reply firstuseauthenticator/firstuseauthenticator.py]: The message is displayed to the console. The user only sees "Invalid username or password" and I believe it comes from [here](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L153).

From what I understand, in order to have our own error message we could define our own *LoginHandler* and overload the *post* method. What do you think? Do you want me to give this a try?

Comment (GeorgianaElena) [reply firstuseauthenticator/firstuseauthenticator.py]: @yuvipanda, I believe a nicer way we could do this is to make the *login_error* a configurable attribute of *LoginHandler* and have it print *Invalid username or password* only if no specific error message is defined. But this implies modifying code in JupyterHub.

**Comment by GeorgianaElena** [comment]
Thanks a lot for your feedback, @yuvipanda :heart: I left some questions in the comments too.
(Do you receive any notification for those? Sorry if you do and I've just flooded you with another one :disappointed: )

**Comment by GeorgianaElena** [comment]
Updated the PR to display the error to the user in the UI not just to the console thanks to @yuvipanda's feedback :confetti_ball: .

When password is shorter than ``min_password_length`` it now looks like this:
* at login:
login_error
* when changing the password:
change_passw_error

File: firstuseauthenticator/firstuseauthenticator.py [hunk 3]
```diff
@@ -10,17 +10,24 @@
...
 
 
 TEMPLATE_DIR = os.path.join(os.path.dirname(__file__), 'templates')
 
 
+class CustomLoginHandler(LoginHandler):
+    custom_login_error = 'Invalid username or password'
+    def _render(self, login_error=None, username=None):
+        return super()._render(self.custom_login_error, username)
+
```
Comment (consideRatio) [inline firstuseauthenticator/firstuseauthenticator.py]: A brief docstring here on the purpose would be useful as Custom is very generic prefix.

I understand from the code that we ignore any passed login_error argument and replace it with custom_login_error. I'm reviewing this in the dark without good insight here, does it make sense to ignore login_error entirely or should the login_error parameter instead have a new default value of custom_login_error?

File: firstuseauthenticator/firstuseauthenticator.py [hunk 4]
```diff
@@ -50,12 +57,18 @@ def _register_template_path(self):
     async def post(self):
         user = self.current_user
         new_password = self.get_body_argument('password', strip=False)
-        self.authenticator.reset_password(user.name, new_password)
+        msg = self.authenticator.reset_password(user.name, new_password)
+
+        if "success" in msg:
+            alert = "success"
```
Comment (consideRatio) [inline firstuseauthenticator/firstuseauthenticator.py]: I'm currently not knowing as to what is contained in `msg`. Is this ok without using `.lower()` first on `msg`? I can imagine someone writing `"Successfully changed the password!"` or similar.

UPDATE: I note below that it comes from a specific message from a function specific to this very class, so it is fine!

File: firstuseauthenticator/firstuseauthenticator.py (hunk 3 above)
Comment (GeorgianaElena) [reply firstuseauthenticator/firstuseauthenticator.py]: I took another look at the code and remembered why I chose to ignore `login_error`.
* In order to customize the `login_error` that gets rendered, I had three options (or at least that's how many I managed to identify):
  1. Override the `post` method of [`class LoginHandler`](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L137) (the place where the error gets rendered in the UI).
  2. Have `login_error` be a `LoginHandler` class member (in JupyterHub repo) and change it when appropriate (right now it's just a parameter passed to the function `_render`). 
  3. Inherit from `LoginHandler` and have a `custom_login_error` that replaces `login_error` in the manner described at option 2.

* The reasons why I chose option 3:
  1. **Option 1** would have meant a lot of duplicated code because the `post` method would have been pretty much the same, except for [this line](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L153).
  2. **Option 2** would have mean messing with the code in JH and I didn't think it was a good idea at that point.
   3. I went with **option 3** because it felt like the most straightforward approach and it didn't need any outside the repo intervention.

Because the default value of `custom_login_error` is set to be [the only value `login_error` can hold](https://github.com/jupyterhub/jupyterhub/blob/f97ef7eaac73e4af4a5444b24fdffe47f388da70/jupyterhub/handlers/login.py#L153) (i.e. *Invalid username or password'* ), `login_error` can be safely ignored without losing any information.
If I would have given `login_error` a new default value of `custom_login_error`, the `post` method call would have overwritten it with *Invalid username or password*. 

What are your thoughts on this @consideRatio?

P.S. A docstring and probably more comments too are definitely needed :sun_with_face:

Comment (consideRatio) [reply firstuseauthenticator/firstuseauthenticator.py]: Ah excellent writedown! So good that you perhaps could reference it by a link in the docstring where you keep it brief about the purpose of the class for example.

You seem to have considered this very thoroughly, given that and I think we could merge this when you are done and release 0.14.0. What do you think?

**Comment by GeorgianaElena** [comment]
Yes, I've tested it manually (and there's also a test in `test_authenticator.py` for it too).
But I'm not 100% sure I've covered all the edge cases or that the solution I chose is the right one :eyes:
Hope it won't break anything.

**Comment by DongchaoChen** [comment]
@GeorgianaElena 

Below is my tljh config
```
$ sudo tljh-config show
users:
  admin:
  - admin
https:
  enabled: true
  letsencrypt:
    email: replace_by_my_email
    domains:
    - replace_by_my_domain.com
auth:
  type: firstuseauthenticator.FirstUseAuthenticator
  FirstUseAuthenticator:
    min_password_length: 7
```
and I am using the latest tljh which will use https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/tljh/installer.py#L226, somehow I still did not see the minimum length on password thing happen when user first time login, which means they can use an empty password. Can you give any suggestion? Many thanks.

**Comment by GeorgianaElena** [comment]
Hey @DongchaoChen :wave: Sorry for the late reply.

The reason why this doesn't work in TLJH is because by default, TLJH has `FirstuseAuthenticator` configured such that the users need to exist in the database before they can login (docs ref [here](https://tljh.jupyter.org/en/latest/howto/auth/firstuse.html#allowing-anyone-to-log-in-to-your-jupyterhub)).

Because the users exist before they login, this means this check will fail and the minimum password requirement won't be enforced for them: https://github.com/jupyterhub/firstuseauthenticator/blob/be01334bbd888a32ad330eb1fe4ba33ac9ece9ed/firstuseauthenticator/firstuseauthenticator.py#L148-L150

Unfortunately I didn't think about this case when adding the min password requirement feature to `FirstuseAuthenticator` :confused:
I will open up an issue about finding a way to enforce this min password requirement for the users that exist in the db, but never logged in and thus, they don't have a password set yet.

Thank you very much @DongchaoChen for reporting this :heart:
